lib/internal/modules/esm/worker.js

```diff
@@ -0,0 +1,89 @@
+'use strict';
+
+const {
+  ReflectApply,
+  SafeWeakMap,
+} = primordials;
+
+let debug = require('internal/util/debuglog').debuglog('esm', (fn) => {
+  debug = fn;
+});
+
+debug('worker for public ESM running');
+
+// Create this WeakMap in js-land because V8 has no C++ API for WeakMap.
+internalBinding('module_wrap').callbackMap = new SafeWeakMap();
+
+const { triggerUncaughtException } = internalBinding('errors');
+const { getOptionValue } = require('internal/options');
+const { ESMLoader } = require('internal/modules/esm/loader');
+const { workerData } = require('worker_threads');
+const { deserialize, serialize } = require('v8');
+const { pathToFileURL } = require('internal/url');
+
+const { commsChannel } = workerData;
+// lock = 0 -> main sleeps
+// lock = 1 -> worker sleeps
+const lock = new Int32Array(commsChannel, 0, 4); // Required by Atomics
+const requestResponseData = new Uint8Array(commsChannel, 4, 2044); // For v8.serialize/deserialize
+
+const publicESMLoader = new ESMLoader();
+
+function loadModulesInIsolation(specifiers, loaders = []) {
+  if (!Array.isArray(specifiers) || specifiers.length === 0) { return; }
+  debug('loading ESM modules', specifiers, 'with loaders', loaders);
+
+  let cwd;
+  try {
+    cwd = process.cwd() + '/';
+  } catch {
+    cwd = '/';
+  }
+
+  // A separate loader instance is necessary to avoid cross-contamination
+  // between internal Node.js and userland. For example, a module with internal
+  // state (such as a counter) should be independent.
+  const internalESMLoader = new ESMLoader();
+  internalESMLoader.addCustomLoaders(loaders);
+
+  // Importation must be handled by internal loader to avoid polluting userland.
+  return internalESMLoader.import(
+    specifiers,
+    pathToFileURL(cwd).href,
+    Object.create(null),
+  );
+}
+
+const types = {
+  'eval': publicESMLoader.eval,
+  'import': publicESMLoader.import,
+  'resolve': publicESMLoader.resolve,
+};
+
+Atomics.store(lock, 0, 1); // Send 'ready' signal to main
+Atomics.notify(lock, 0); // Notify main of signal
+
+(async () => {
+  const customLoaders = getOptionValue('--experimental-loader');
+  const loaders = await loadModulesInIsolation(customLoaders);
+
+  // Hooks must then be added to external/public loader
+  // (so they're triggered in userland).
+  publicESMLoader.addCustomLoaders(loaders);
+
+  while (true) { // event loop
+    debug('blocking worker thread until main thread is ready');
+    Atomics.wait(lock, 0, 1); // This pauses the while loop
```

Comment:
Any reason not to use waitAsync here?

Reply:
Are you talking about `async`/`await`? Yes, the purpose of this is to force the main thread to behave synchronously whilst the worker thread remains async.

Reply:
I think Robert is talking about [`Atomics.waitAsync`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/waitAsync). IIUC this method doesn't block the main thread, which is what we are trying to do here.

Reply:
Ah 🤦‍♂️ yes and yes

Reply:
I mean this is the worker thread? Does that need to be blocked?

Reply:
It has nothing to do, so it would otherwise loop continuously for no reason. It's "listening" for a "go" signal via the shared data buffer, and this can't use postMessage et al (it creates a cart-before-the-horse scenario), so I can't think of another way to handle it.

Is there a downside to blocking the worker thread when it has nothing to do?